File: go/analyze/rules_tx.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package analyze

import (
	"context"
	"regexp"
	"strings"

	keystonev1alpha1 "github.com/dogukanturhal/keystone/api/v1alpha1"
)

// Phase A2 — transaction and session-state rules.
//
// Keystone's runner wraps every migration file in an implicit
// transaction (see internal/migration/runner.go). That invariant is
// load-bearing — on failure the runner rolls back and nothing leaks.
// These rules flag session-state manipulation that breaks that
// invariant (committing / rolling back mid-migration) or bypasses
// trigger-based integrity (session_replication_role, SET CONSTRAINTS
// DEFERRED as an escape hatch).

// -- Rule: no-commit-in-migration --------------------------------------

// NoCommitInMigration flags bare COMMIT inside a migration file. The
// runner owns transaction control; an author-supplied COMMIT splits a
// single logical migration into multiple physical transactions, which
// silently prevents rollback on failure.
type NoCommitInMigration struct{}

func (NoCommitInMigration) ID() string { return "no-commit-in-migration" }
func (NoCommitInMigration) Description() string {
	return "COMMIT inside a migration breaks the runner's rollback invariant"
}

// Match COMMIT on its own line (with optional trailing semicolon).
// Avoid matching 2PC variants like `COMMIT PREPARED 'xid'` which have
// different semantics.
var reCommit = regexp.MustCompile(`(?i)^\s*COMMIT\s*;?\s*(--.*)?$`)

func (a *NoCommitInMigration) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		for i, line := range strings.Split(f.Body, "\n") {
			if !reCommit.MatchString(line) {
				continue
			}
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelError,
				File:     f.Name,
				Line:     int32(i + 1),
				Message: "COMMIT inside the migration body splits it into multiple physical " +
					"transactions. The runner wraps each file in its own transaction; " +
					"an author-supplied COMMIT makes partial failures non-rollbackable.",
			})
		}
	}
	return out, nil
}

// -- Rule: no-rollback-in-migration ------------------------------------

// NoRollbackInMigration mirrors NoCommitInMigration for ROLLBACK. The
// runner rolls back automatically on error; an explicit ROLLBACK inside
// the migration is almost always a leftover debug statement.
type NoRollbackInMigration struct{}

func (NoRollbackInMigration) ID() string { return "no-rollback-in-migration" }
func (NoRollbackInMigration) Description() string {
	return "ROLLBACK inside a migration is almost always a leftover debug statement"
}

var reRollback = regexp.MustCompile(`(?i)^\s*ROLLBACK\s*;?\s*(--.*)?$`)

func (a *NoRollbackInMigration) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		for i, line := range strings.Split(f.Body, "\n") {
			if !reRollback.MatchString(line) {
				continue
			}
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelError,
				File:     f.Name,
				Line:     int32(i + 1),
				Message: "ROLLBACK inside the migration body is a debug remnant. The runner " +
					"rolls back automatically when a statement errors; explicit ROLLBACK " +
					"discards work the author apparently meant to commit.",
			})
		}
	}
	return out, nil
}

// -- Rule: no-set-session-replication-role -----------------------------

// NoSetSessionReplicationRole flags the classic FK-bypass trick
// `SET session_replication_role = replica`. Under that setting,
// user-defined triggers (including FK triggers) don't fire, so an
// INSERT can plant an orphan row that violates every downstream
// invariant.
type NoSetSessionReplicationRole struct{}

func (NoSetSessionReplicationRole) ID() string { return "no-set-session-replication-role" }
func (NoSetSessionReplicationRole) Description() string {
	return "SET session_replication_role bypasses FK and user triggers"
}

var reSessionReplicationRole = regexp.MustCompile(`(?i)\bSET\s+(SESSION\s+|LOCAL\s+)?session_replication_role\b`)

func (a *NoSetSessionReplicationRole) Check(_ context.Context, m *Migration) ([]Finding, error) {
	return statementMatches(m, reSessionReplicationRole, keystonev1alpha1.LintLevelError,
		a.ID(),
		"SET session_replication_role silences every user-defined trigger, "+
			"including FK enforcement. Orphan rows planted this way survive the migration "+
			"and break every downstream invariant. Use CHECK NOT VALID + VALIDATE CONSTRAINT "+
			"if you need a deferred constraint check instead."), nil
}

// -- Rule: no-set-constraints-deferred ---------------------------------

// NoSetConstraintsDeferred flags SET CONSTRAINTS ALL DEFERRED (or its
// named form). Deferring the FK check to COMMIT time is a legitimate
// feature for circular FK inserts, but the blast-radius form
// `SET CONSTRAINTS ALL DEFERRED` is overkill in 99% of cases — it
// defers every constraint, including ones that should fire immediately
// for tight feedback.
type NoSetConstraintsDeferred struct{}

func (NoSetConstraintsDeferred) ID() string { return "no-set-constraints-deferred" }
func (NoSetConstraintsDeferred) Description() string {
	return "SET CONSTRAINTS ALL DEFERRED delays every FK/CHECK validation to COMMIT"
}

var reSetConstraintsDeferred = regexp.MustCompile(`(?i)\bSET\s+CONSTRAINTS\s+ALL\s+DEFERRED\b`)

func (a *NoSetConstraintsDeferred) Check(_ context.Context, m *Migration) ([]Finding, error) {
	return statementMatches(m, reSetConstraintsDeferred, keystonev1alpha1.LintLevelWarning,
		a.ID(),
		"SET CONSTRAINTS ALL DEFERRED postpones every FK/CHECK to COMMIT. "+
			"For circular inserts, defer only the specific constraint: "+
			"`SET CONSTRAINTS <name> DEFERRED`. Wholesale deferral hides errors "+
			"that should fire immediately and makes debugging harder."), nil
}

File: rust/keystone-sdk/src/analyze/rules_tx.rs
// SPDX-License-Identifier: Apache-2.0

//! Phase A2 — transaction / session-state safety. Ported from `rules_tx.go`.

use std::sync::LazyLock;

use regex::Regex;

use super::{finding, statement_matches, Analyzer, Finding, LintLevel, Migration};

macro_rules! re {
    ($name:ident, $pat:expr) => {
        static $name: LazyLock<Regex> = LazyLock::new(|| Regex::new($pat).expect($pat));
    };
}

re!(RE_COMMIT, r"(?i)^\s*COMMIT\s*;?\s*(--.*)?$");
re!(RE_ROLLBACK, r"(?i)^\s*ROLLBACK\s*;?\s*(--.*)?$");
re!(
    RE_SESSION_REPLICATION_ROLE,
    r"(?i)\bSET\s+(SESSION\s+|LOCAL\s+)?session_replication_role\b"
);
re!(
    RE_SET_CONSTRAINTS_DEFERRED,
    r"(?i)\bSET\s+CONSTRAINTS\s+ALL\s+DEFERRED\b"
);

/// no-commit-in-migration
pub struct NoCommitInMigration;
impl Analyzer for NoCommitInMigration {
    fn id(&self) -> &'static str {
        "no-commit-in-migration"
    }
    fn description(&self) -> &'static str {
        "COMMIT inside a migration breaks the runner's rollback invariant"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            for (i, line) in f.body.split('\n').enumerate() {
                if !RE_COMMIT.is_match(line) {
                    continue;
                }
                out.push(finding(
                    self.id(),
                    LintLevel::Error,
                    &f.name,
                    (i + 1) as i32,
                    "COMMIT inside the migration body splits it into multiple physical \
                     transactions. The runner wraps each file in its own transaction; \
                     an author-supplied COMMIT makes partial failures non-rollbackable.",
                ));
            }
        }
        out
    }
}

/// no-rollback-in-migration
pub struct NoRollbackInMigration;
impl Analyzer for NoRollbackInMigration {
    fn id(&self) -> &'static str {
        "no-rollback-in-migration"
    }
    fn description(&self) -> &'static str {
        "ROLLBACK inside a migration is almost always a leftover debug statement"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            for (i, line) in f.body.split('\n').enumerate() {
                if !RE_ROLLBACK.is_match(line) {
                    continue;
                }
                out.push(finding(
                    self.id(),
                    LintLevel::Error,
                    &f.name,
                    (i + 1) as i32,
                    "ROLLBACK inside the migration body is a debug remnant. The runner \
                     rolls back automatically when a statement errors; explicit ROLLBACK \
                     discards work the author apparently meant to commit.",
                ));
            }
        }
        out
    }
}

/// no-set-session-replication-role
pub struct NoSetSessionReplicationRole;
impl Analyzer for NoSetSessionReplicationRole {
    fn id(&self) -> &'static str {
        "no-set-session-replication-role"
    }
    fn description(&self) -> &'static str {
        "SET session_replication_role bypasses FK and user triggers"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        statement_matches(
            m,
            &RE_SESSION_REPLICATION_ROLE,
            LintLevel::Error,
            self.id(),
            "SET session_replication_role silences every user-defined trigger, \
             including FK enforcement. Orphan rows planted this way survive the migration \
             and break every downstream invariant. Use CHECK NOT VALID + VALIDATE CONSTRAINT \
             if you need a deferred constraint check instead.",
        )
    }
}

/// no-set-constraints-deferred
pub struct NoSetConstraintsDeferred;
impl Analyzer for NoSetConstraintsDeferred {
    fn id(&self) -> &'static str {
        "no-set-constraints-deferred"
    }
    fn description(&self) -> &'static str {
        "SET CONSTRAINTS ALL DEFERRED delays every FK/CHECK validation to COMMIT"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        statement_matches(
            m,
            &RE_SET_CONSTRAINTS_DEFERRED,
            LintLevel::Warning,
            self.id(),
            "SET CONSTRAINTS ALL DEFERRED postpones every FK/CHECK to COMMIT. \
             For circular inserts, defer only the specific constraint: \
             `SET CONSTRAINTS <name> DEFERRED`. Wholesale deferral hides errors \
             that should fire immediately and makes debugging harder.",
        )
    }
}

#[cfg(test)]
mod tests {
    use super::super::{FileBody, Migration};
    use super::*;

    fn mig(body: &str) -> Migration {
        Migration {
            files: vec![FileBody {
                name: "001.sql".into(),
                body: body.into(),
            }],
            ..Default::default()
        }
    }

    #[test]
    fn commit_and_rollback() {
        assert_eq!(NoCommitInMigration.check(&mig("COMMIT;")).len(), 1);
        assert_eq!(NoCommitInMigration.check(&mig("  commit")).len(), 1);
        // COMMIT PREPARED is a 2PC variant — must NOT match.
        assert_eq!(
            NoCommitInMigration
                .check(&mig("COMMIT PREPARED 'x';"))
                .len(),
            0
        );
        assert_eq!(NoRollbackInMigration.check(&mig("ROLLBACK;")).len(), 1);
    }

    #[test]
    fn session_replication_role_and_constraints() {
        assert_eq!(
            NoSetSessionReplicationRole
                .check(&mig("SET session_replication_role = replica;"))
                .len(),
            1
        );
        assert_eq!(
            NoSetConstraintsDeferred
                .check(&mig("SET CONSTRAINTS ALL DEFERRED;"))
                .len(),
            1
        );
    }
}
